File: benchmarking/python_bench/bench_write.py
#!/usr/bin/env python3
"""MCAP write benchmark for Python."""

import os
import resource
import sys
import time
import zlib

# Shared payload blob parameters; must match gen_blob.py and the other
# language benches. Message i's payload is the window of the blob starting
# at (i * BLOB_STRIDE) % BLOB_WINDOW_SPAN, so all implementations feed
# identical bytes to their writers.
BLOB_SIZE = 16777216
BLOB_MAX_PAYLOAD = 524288
BLOB_WINDOW_SPAN = BLOB_SIZE - BLOB_MAX_PAYLOAD
BLOB_STRIDE = 7919


def payload_offset(msg_index):
    return (msg_index * BLOB_STRIDE) % BLOB_WINDOW_SPAN


def load_blob(path):
    with open(path, "rb") as f:
        blob = f.read()
    if len(blob) != BLOB_SIZE:
        print(f"Blob file {path} is not exactly {BLOB_SIZE} bytes", file=sys.stderr)
        sys.exit(1)
    return memoryview(blob)


def peak_rss_kb() -> int:
    """Peak RSS in KB. ru_maxrss is KB on Linux but bytes on macOS."""
    rss = resource.getrusage(resource.RUSAGE_SELF).ru_maxrss
    if sys.platform == "darwin":
        rss //= 1024
    return rss


def main():
    if len(sys.argv) != 6:
        print(
            f"Usage: {sys.argv[0]} <output_file> <mode> <num_messages> <payload_size> <blob_file>",
            file=sys.stderr,
        )
        return 1

    blob = load_blob(sys.argv[5])

    filename = sys.argv[1]
    mode = sys.argv[2]
    mixed = sys.argv[4] == "mixed"

    if mixed:
        num_messages = 3750
        payload_size_str = "mixed"
    else:
        num_messages = int(sys.argv[3])
        payload_size_str = sys.argv[4]
        if int(sys.argv[4]) > BLOB_MAX_PAYLOAD:
            print(f"payload_size must be <= {BLOB_MAX_PAYLOAD}", file=sys.stderr)
            return 1

    from mcap.writer import Writer, CompressionType

    with open(filename, "wb") as f:
        if mode == "unchunked":
            writer = Writer(f, use_chunking=False)
        elif mode == "chunked":
            writer = Writer(f, compression=CompressionType.NONE, chunk_size=786432)
        elif mode == "zstd":
            writer = Writer(f, compression=CompressionType.ZSTD, chunk_size=786432)
        elif mode == "lz4":
            writer = Writer(f, compression=CompressionType.LZ4, chunk_size=786432)
        else:
            print(f"Unknown mode: {mode}", file=sys.stderr)
            return 1

        writer.start(profile="bench", library="py-bench")

        if mixed:
            # Channel definitions: (topic, schema_name, base_payload_size, period_ns, count)
            channel_defs = [
                ("/imu", "IMU", 96, 5_000_000, 2000),
                ("/odom", "Odometry", 296, 20_000_000, 500),
                ("/tf", "TFMessage", None, 10_000_000, 1000),
                ("/lidar", "PointCloud2", 230_400, 100_000_000, 100),
                ("/camera/compressed", "CompressedImage", 524_288, 66_666_667, 150),
            ]

            tf_payload_cycle = [80, 160, 320, 800, 1600]

            schema_ids = []
            channel_ids = []
            for topic, schema_name, _, _, _ in channel_defs:
                sid = writer.register_schema(
                    name=schema_name,
                    encoding="jsonschema",
                    data=b'{"type":"object"}',
                )
                cid = writer.register_channel(
                    topic=topic,
                    message_encoding="json",
                    schema_id=sid,
                )
                schema_ids.append(sid)
                channel_ids.append(cid)

            # Pre-generate the message schedule sorted by (timestamp, channel_index)
            schedule = []
            for ch_idx, (_, _, _, period_ns, count) in enumerate(channel_defs):
                for msg_i in range(count):
                    ts = msg_i * period_ns
                    schedule.append((ts, ch_idx, msg_i))
            schedule.sort(key=lambda x: (x[0], x[1]))

            # Not timed: CRC of the payload stream for cross-language verification
            payload_crc = 0
            for i, (ts, ch_idx, msg_i) in enumerate(schedule):
                if ch_idx == 2:  # /tf
                    size = tf_payload_cycle[msg_i % len(tf_payload_cycle)]
                else:
                    size = channel_defs[ch_idx][2]
                off = payload_offset(i)
                payload_crc = zlib.crc32(blob[off : off + size], payload_crc)

            # Per-channel sequence counters
            seq = [0] * len(channel_defs)

            # Time the message-writing loop + finish + flush to the OS
            t_start = time.perf_counter_ns()

            for i, (ts, ch_idx, msg_i) in enumerate(schedule):
                if ch_idx == 2:  # /tf
                    size = tf_payload_cycle[msg_i % len(tf_payload_cycle)]
                else:
                    size = channel_defs[ch_idx][2]
                off = payload_offset(i)
                payload = blob[off : off + size]
                writer.add_message(
                    channel_id=channel_ids[ch_idx],
                    sequence=seq[ch_idx],
                    log_time=ts,
                    publish_time=ts,
                    data=payload,
                )
                seq[ch_idx] += 1

            writer.finish()
            f.flush()
            t_end = time.perf_counter_ns()
        else:
            payload_size = int(sys.argv[4])

            schema_id = writer.register_schema(
                name="BenchMsg",
                encoding="jsonschema",
                data=b'{"type":"object"}',
            )

            channel_id = writer.register_channel(
                topic="/bench",
                message_encoding="json",
                schema_id=schema_id,
            )

            # Not timed: CRC of the payload stream for cross-language verification
            payload_crc = 0
            for i in range(num_messages):
                off = payload_offset(i)
                payload_crc = zlib.crc32(blob[off : off + payload_size], payload_crc)

            # Time the message-writing loop + finish + flush to the OS
            t_start = time.perf_counter_ns()

            for i in range(num_messages):
                log_time = i * 1000
                off = payload_offset(i)
                writer.add_message(
                    channel_id=channel_id,
                    sequence=i,
                    log_time=log_time,
                    publish_time=log_time,
                    data=blob[off : off + payload_size],
                )

            writer.finish()
            f.flush()
            t_end = time.perf_counter_ns()

    elapsed_ns = t_end - t_start
    wall_sec = elapsed_ns / 1e9
    file_size = os.path.getsize(filename)

    rss_kb = peak_rss_kb()

    # TSV output: op lang mode num_msgs payload_size file_size elapsed_ns wall_sec peak_rss_kb payload_crc32
    print(
        f"write\tpython\t{mode}\t{num_messages}\t{payload_size_str}\t{file_size}\t{elapsed_ns}\t{wall_sec:.6f}\t{rss_kb}\t{payload_crc}"
    )

    return 0


if __name__ == "__main__":
    sys.exit(main())

File: benchmarking/typescript_bench/bench_write.ts
#!/usr/bin/env -S npx tsx
/**
 * MCAP write benchmark for TypeScript.
 */

import { open, readFile, stat } from "fs/promises";
import { crc32 } from "zlib";

import { McapWriter } from "../../typescript/core/src/index.ts";
import { FileHandleWritable } from "../../typescript/nodejs/src/index.ts";

// Shared payload blob parameters; must match gen_blob.py and the other
// language benches. Message i's payload is the window of the blob starting
// at (i * BLOB_STRIDE) % BLOB_WINDOW_SPAN, so all implementations feed
// identical bytes to their writers.
const BLOB_SIZE = 16777216;
const BLOB_MAX_PAYLOAD = 524288;
const BLOB_WINDOW_SPAN = BLOB_SIZE - BLOB_MAX_PAYLOAD;
const BLOB_STRIDE = 7919;

function payloadOffset(msgIndex: number): number {
  return (msgIndex * BLOB_STRIDE) % BLOB_WINDOW_SPAN;
}

async function main(): Promise<number> {
  if (process.argv.length !== 7) {
    process.stderr.write(
      `Usage: ${process.argv[1]} <output_file> <mode> <num_messages> <payload_size> <blob_file>\n`,
    );
    return 1;
  }

  const blob: Uint8Array = await readFile(process.argv[6]!);
  if (blob.byteLength !== BLOB_SIZE) {
    process.stderr.write(
      `Blob file ${process.argv[6]!} is not exactly ${BLOB_SIZE} bytes\n`,
    );
    return 1;
  }

  const filename = process.argv[2]!;
  const mode = process.argv[3]!;
  const isMixed = process.argv[5] === "mixed";
  const numMessages = isMixed ? 3750 : parseInt(process.argv[4]!, 10);
  const payloadSize = isMixed ? 0 : parseInt(process.argv[5]!, 10);

  if (!isMixed && (numMessages <= 0 || payloadSize <= 0)) {
    process.stderr.write("num_messages and payload_size must be positive\n");
    return 1;
  }
  if (!isMixed && payloadSize > BLOB_MAX_PAYLOAD) {
    process.stderr.write(`payload_size must be <= ${BLOB_MAX_PAYLOAD}\n`);
    return 1;
  }

  let useChunks = true;
  let compressChunk:
    | ((data: Uint8Array) => {
        compression: string;
        compressedData: Uint8Array;
      })
    | undefined;

  if (mode === "unchunked") {
    useChunks = false;
  } else if (mode === "chunked") {
    // chunked, no compression
  } else if (mode === "zstd") {
    const zstdMod = await import("@foxglove/wasm-zstd");
    const zstd = zstdMod.default;
    await zstd.isLoaded;
    compressChunk = (data: Uint8Array) => ({
      compression: "zstd",
      compressedData: new Uint8Array(zstd.compress(data)),
    });
  } else if (mode === "lz4") {
    process.stderr.write(
      "LZ4 compression is not available in the TypeScript MCAP library\n",
    );
    return 1;
  } else {
    process.stderr.write(`Unknown mode: ${mode}\n`);
    return 1;
  }

  const fileHandle = await open(filename, "w");
  const writer = new McapWriter({
    writable: new FileHandleWritable(fileHandle),
    useChunks,
    chunkSize: 786432,
    compressChunk,
  });

  await writer.start({ profile: "bench", library: "ts-bench" });

  let tStart: bigint;
  let tEnd: bigint;
  let payloadCrc = 0;

  if (isMixed) {
    // Mixed-payload mode: simulate a 10-second robot recording.
    const channelDefs = [
      {
        topic: "/imu",
        schema: "IMU",
        sizes: [96],
        periodNs: 5_000_000n,
        count: 2000,
      },
      {
        topic: "/odom",
        schema: "Odometry",
        sizes: [296],
        periodNs: 20_000_000n,
        count: 500,
      },
      {
        topic: "/tf",
        schema: "TFMessage",
        sizes: [80, 160, 320, 800, 1600],
        periodNs: 10_000_000n,
        count: 1000,
      },
      {
        topic: "/lidar",
        schema: "PointCloud2",
        sizes: [230400],
        periodNs: 100_000_000n,
        count: 100,
      },
      {
        topic: "/camera/compressed",
        schema: "CompressedImage",
        sizes: [524288],
        periodNs: 66_666_667n,
        count: 150,
      },
    ];

    const schemaData = new TextEncoder().encode('{"type":"object"}');
    const channelIds: number[] = [];
    for (const def of channelDefs) {
      const sid = await writer.registerSchema({
        name: def.schema,
        encoding: "jsonschema",
        data: schemaData,
      });
      const cid = await writer.registerChannel({
        topic: def.topic,
        schemaId: sid,
        messageEncoding: "json",
        metadata: new Map(),
      });
      channelIds.push(cid);
    }

    // Pre-generate sorted message schedule.
    const schedule: { timestamp: bigint; channelIndex: number }[] = [];
    for (let ci = 0; ci < channelDefs.length; ci++) {
      const def = channelDefs[ci]!;
      for (let m = 0; m < def.count; m++) {
        schedule.push({
          timestamp: BigInt(m) * def.periodNs,
          channelIndex: ci,
        });
      }
    }
    schedule.sort((a, b) => {
      if (a.timestamp < b.timestamp) return -1;
      if (a.timestamp > b.timestamp) return 1;
      return a.channelIndex - b.channelIndex;
    });

    // Not timed: CRC of the payload stream for cross-language verification.
    {
      const crcSeq = new Array(channelDefs.length).fill(0) as number[];
      for (let i = 0; i < schedule.length; i++) {
        const ci = schedule[i]!.channelIndex;
        const def = channelDefs[ci]!;
        const seq = crcSeq[ci]!;
        crcSeq[ci] = seq + 1;
        const sz = def.sizes[seq % def.sizes.length]!;
        const off = payloadOffset(i);
        payloadCrc = crc32(blob.subarray(off, off + sz), payloadCrc);
      }
    }

    // Per-channel sequence counters for payload size cycling.
    const chanSeq = new Array(channelDefs.length).fill(0) as number[];

    // Time the message-writing loop + end.
    tStart = process.hrtime.bigint();

    for (let i = 0; i < schedule.length; i++) {
      const msg = schedule[i]!;
      const ci = msg.channelIndex;
      const def = channelDefs[ci]!;
      const seq = chanSeq[ci]!;
      chanSeq[ci] = seq + 1;
      const sz = def.sizes[seq % def.sizes.length]!;
      const off = payloadOffset(i);
      const data = blob.subarray(off, off + sz);
      await writer.addMessage({
        channelId: channelIds[ci]!,
        sequence: seq,
        logTime: msg.timestamp,
        publishTime: msg.timestamp,
        data,
      });
    }

    await writer.end();
    tEnd = process.hrtime.bigint();
    await fileHandle.close();
  } else {
    // Fixed-payload mode.
    const schemaId = await writer.registerSchema({
      name: "BenchMsg",
      encoding: "jsonschema",
      data: new TextEncoder().encode('{"type":"object"}'),
    });

    const channelId = await writer.registerChannel({
      topic: "/bench",
      schemaId,
      messageEncoding: "json",
      metadata: new Map(),
    });

    // Not timed: CRC of the payload stream for cross-language verification.
    for (let i = 0; i < numMessages; i++) {
      const off = payloadOffset(i);
      payloadCrc = crc32(blob.subarray(off, off + payloadSize), payloadCrc);
    }

    // Time the message-writing loop + end
    tStart = process.hrtime.bigint();

    for (let i = 0; i < numMessages; i++) {
      const logTime = BigInt(i) * 1000n;
      const off = payloadOffset(i);
      await writer.addMessage({
        channelId,
        sequence: i,
        logTime,
        publishTime: logTime,
        data: blob.subarray(off, off + payloadSize),
      });
    }

    await writer.end();
    tEnd = process.hrtime.bigint();
    await fileHandle.close();
  }

  const elapsedNs = tEnd - tStart;
  const wallSec = Number(elapsedNs) / 1e9;
  const fileSize = (await stat(filename)).size;

  const peakRssKb = process.resourceUsage().maxRSS;

  // TSV output: op lang mode num_msgs payload_size file_size elapsed_ns wall_sec peak_rss_kb payload_crc32
  const payloadSizeStr = isMixed ? "mixed" : String(payloadSize);
  process.stdout.write(
    `write\ttypescript\t${mode}\t${numMessages}\t${payloadSizeStr}\t${fileSize}\t${elapsedNs}\t${wallSec.toFixed(
      6,
    )}\t${peakRssKb}\t${payloadCrc}\n`,
  );

  return 0;
}

main().then((code) => process.exit(code));
